src/unix/stream.c

```diff
@@ -1311,9 +1311,10 @@ static void uv__stream_io(uv_loop_t* loop, uv__io_t* w, unsigned int events) {
 
   assert(uv__stream_fd(stream) >= 0);
 
-  /* Ignore POLLHUP here. Even if it's set, there may still be data to read. */
-  if (events & (POLLIN | POLLERR | POLLHUP))
+  if (events & (POLLIN | POLLERR))
     uv__read(stream);
+  else
+    stream->flags |= UV_HANDLE_READ_PARTIAL;
```

Comment:
I suspect this causes a visible change in programs; they'll see UV_EOF instead of e.g. UV_EPIPE. Can you try running Node's test suite with this change?

Reply:
Good idea

A possibly bigger semi-related question I noticed is that in kqueue, we appear to interpret EVFILT_WRITE+EV_EOF (aka reader-disconnected) as UV__POLLRDHUP / UV_DISCONNECTED (aka writer-disconnected), if I am reading the kqueue code and documentation correctly. It is probably difficult for there to be enough data in the read queue, and rare for people to use `shutdown(SHUT_RD)`, but this PR would cause us to truncate those reads due that incorrect flag (which already perhaps gets incorrectly exposed to uv_poll users however). I will make a PR.